src/index.js

```diff
@@ -1,160 +1,333 @@
-      const subopts = topics.map((topic) => {
-        return {
-          subscribe: false,
-          topicCID: topic
-        }
-      })
-      const rpc = pb.rpc.RPC.encode({
-        subscriptions: subopts
+  /**
+   * Publish messages to the given topics.
+   *
+   * @param {Array<string>|string} topics
+   * @param {Array<any>|any} messages
+   * @returns {undefined}
+   *
+   */
+  publish (topics, messages) {
+    log('publish', topics, messages)
+
+    topics = ensureArray(topics)
+    messages = ensureArray(messages)
+
+    // Emit to self if I'm interested
+    topics
+      .filter((topic) => this.subscriptions.has(topic))
+      .forEach((topic) => {
+        messages.forEach((message) => {
+          this.emit(topic, message)
+        })
       })
 
-      peer.stream.write(rpc)
+    const from = this.libp2p.peerInfo.id.toB58String()
+
+    const buildMessage = (msg) => {
+      const seqno = utils.randomSeqno()
+      this.cache.put(utils.msgId(from, seqno))
+
+      return {
+        from: from,
+        data: msg,
+        seqno: new Buffer(seqno),
+        topicCIDs: topics
+      }
+    }
+
+    // send to all the other peers
+    this.peers.forEach((peer) => {
+      if (!utils.anyMatch(peer.topics, topics)) {
+        return
+      }
+
+      peer.write(pb.rpc.RPC.encode({
+        msgs: messages.map(buildMessage)
+      }))
+
+      log('publish msgs on topics', topics, peer.info.id.toB58String())
     })
   }
 
-  this.getPeerSet = () => {
-    return peerSet
-  }
+  /**
+   * Subscribe to the given topic(s).
+   *
+   * @param {Array<string>|string} topics
+   * @returns {undefined}
+   */
+  subscribe (topics) {
+    topics = ensureArray(topics)
 
-  this.getSubscriptions = () => {
-    return subscriptions
-  }
+    topics.forEach((topic) => {
+      this.subscriptions.add(topic)
+    })
 
-  function onStart () {
-    const connectedPeers = libp2pNode.peerBook.getAll()
-    _values(connectedPeers).forEach(dial)
+    const buildSubscription = (topic) => {
+      return {
+        subscribe: true,
+        topicCID: topic
+      }
+    }
+
+    this.peers.forEach((peer) => {
+      peer.write(pb.rpc.RPC.encode({
+        subscriptions: topics.map(buildSubscription)
+      }))
+    })
   }
-  onStart()
 
-  // speed up any new peer that comes in my way
-  libp2pNode.swarm.on('peer-mux-established', dial)
+  /**
+   * Unsubscribe from the given topic(s).
+   *
+   * @param {Array<string>|string} topics
+   * @returns {undefined}
+   */
+  unsubscribe (topics) {
+    topics = ensureArray(topics)
+
+    topics.forEach((topic) => {
+      this.subscriptions.delete(topic)
+    })
+
+    const buildUnsubscription = (topic) => {
+      return {
+        subscribe: false,
+        topicCID: topic
+      }
+    }
+
+    this.peers.forEach((peer) => {
+      peer.write(pb.rpc.RPC.encode({
+        subscriptions: topics.map(buildUnsubscription)
+      }))
+    })
+  }
 }
+
+module.exports = PubSubGossip
```

Comment:
Yeah about that, I moved this in here because I was seeing issues with references to `this.peers` not being properly updated when passing them around to another file. As soon as I moved those methods into the `class` things were fine :/ 

I will see if I can pull something out again.